File: packages/java-sdk/src/main/java/com/turbodocx/models/TemplateVariable.java
package com.turbodocx.models;

import java.util.List;
import java.util.Map;

/**
 * Variable configuration for template generation
 * <p>
 * Supports both simple text replacement and advanced templating with Angular-like expressions
 */
public class TemplateVariable {
    private String placeholder;

    private String name;

    private Object value;

    private String text;

    private String mimeType;

    private Boolean usesAdvancedTemplatingEngine;

    private Boolean nestedInAdvancedTemplatingEngine;

    private Boolean allowRichTextInjection;

    private String description;

    private Boolean defaultValue;

    private List<TemplateVariable> subvariables;

    // Constructors
    public TemplateVariable() {
    }

    public TemplateVariable(String placeholder, Object value) {
        this.placeholder = placeholder;
        this.value = value;
    }

    // Getters and Setters
    public String getPlaceholder() {
        return placeholder;
    }

    public void setPlaceholder(String placeholder) {
        this.placeholder = placeholder;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Object getValue() {
        return value;
    }

    public void setValue(Object value) {
        this.value = value;
    }

    public String getText() {
        return text;
    }

    public void setText(String text) {
        this.text = text;
    }

    public String getMimeType() {
        return mimeType;
    }

    public void setMimeType(String mimeType) {
        this.mimeType = mimeType;
    }

    public void setMimeType(VariableMimeType mimeType) {
        this.mimeType = mimeType.getValue();
    }

    public Boolean getUsesAdvancedTemplatingEngine() {
        return usesAdvancedTemplatingEngine;
    }

    public void setUsesAdvancedTemplatingEngine(Boolean usesAdvancedTemplatingEngine) {
        this.usesAdvancedTemplatingEngine = usesAdvancedTemplatingEngine;
    }

    public Boolean getNestedInAdvancedTemplatingEngine() {
        return nestedInAdvancedTemplatingEngine;
    }

    public void setNestedInAdvancedTemplatingEngine(Boolean nestedInAdvancedTemplatingEngine) {
        this.nestedInAdvancedTemplatingEngine = nestedInAdvancedTemplatingEngine;
    }

    public Boolean getAllowRichTextInjection() {
        return allowRichTextInjection;
    }

    public void setAllowRichTextInjection(Boolean allowRichTextInjection) {
        this.allowRichTextInjection = allowRichTextInjection;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public Boolean getDefaultValue() {
        return defaultValue;
    }

    public void setDefaultValue(Boolean defaultValue) {
        this.defaultValue = defaultValue;
    }

    public List<TemplateVariable> getSubvariables() {
        return subvariables;
    }

    public void setSubvariables(List<TemplateVariable> subvariables) {
        this.subvariables = subvariables;
    }

    // Builder pattern
    public static class Builder {
        private final TemplateVariable variable = new TemplateVariable();

        public Builder placeholder(String placeholder) {
            variable.placeholder = placeholder;
            return this;
        }

        public Builder name(String name) {
            variable.name = name;
            return this;
        }

        public Builder value(Object value) {
            variable.value = value;
            return this;
        }

        public Builder text(String text) {
            variable.text = text;
            return this;
        }

        public Builder mimeType(String mimeType) {
            variable.mimeType = mimeType;
            return this;
        }

        public Builder mimeType(VariableMimeType mimeType) {
            variable.mimeType = mimeType.getValue();
            return this;
        }

        public Builder usesAdvancedTemplatingEngine(Boolean usesAdvancedTemplatingEngine) {
            variable.usesAdvancedTemplatingEngine = usesAdvancedTemplatingEngine;
            return this;
        }

        public Builder nestedInAdvancedTemplatingEngine(Boolean nestedInAdvancedTemplatingEngine) {
            variable.nestedInAdvancedTemplatingEngine = nestedInAdvancedTemplatingEngine;
            return this;
        }

        public Builder allowRichTextInjection(Boolean allowRichTextInjection) {
            variable.allowRichTextInjection = allowRichTextInjection;
            return this;
        }

        public Builder description(String description) {
            variable.description = description;
            return this;
        }

        public Builder defaultValue(Boolean defaultValue) {
            variable.defaultValue = defaultValue;
            return this;
        }

        public Builder subvariables(List<TemplateVariable> subvariables) {
            variable.subvariables = subvariables;
            return this;
        }

        public TemplateVariable build() {
            if (variable.placeholder == null || variable.placeholder.isEmpty()) {
                throw new IllegalStateException("placeholder must be set");
            }
            if (variable.name == null || variable.name.isEmpty()) {
                throw new IllegalStateException("name must be set");
            }
            if (variable.mimeType == null || variable.mimeType.isEmpty()) {
                throw new IllegalStateException("mimeType must be set");
            }
            return variable;
        }
    }

    public static Builder builder() {
        return new Builder();
    }

    // Helper factory methods

    /**
     * Creates a simple text variable
     * @param placeholder The variable placeholder (e.g., "{customer_name}")
     * @param name The variable name
     * @param value The value to substitute
     * @param mimeType The mime type (TEXT or HTML)
     * @throws IllegalArgumentException if any required parameter is missing or invalid
     */
    public static TemplateVariable simple(String placeholder, String name, Object value, VariableMimeType mimeType) {
        if (placeholder == null || placeholder.isEmpty()) {
            throw new IllegalArgumentException("placeholder is required");
        }
        if (name == null || name.isEmpty()) {
            throw new IllegalArgumentException("name is required");
        }
        if (mimeType == null) {
            throw new IllegalArgumentException("mimeType is required");
        }
        if (mimeType != VariableMimeType.TEXT && mimeType != VariableMimeType.HTML) {
            throw new IllegalArgumentException("mimeType must be TEXT or HTML");
        }
        return builder()
                .placeholder(placeholder)
                .name(name)
                .value(value)
                .mimeType(mimeType)
                .build();
    }

    /**
     * Creates an advanced engine variable (for nested objects, complex data)
     * @param placeholder The variable placeholder (e.g., "{user}")
     * @param name The variable name
     * @param value The nested object value
     * @throws IllegalArgumentException if any required parameter is missing
     */
    public static TemplateVariable advancedEngine(String placeholder, String name, Map<String, Object> value) {
        if (placeholder == null || placeholder.isEmpty()) {
            throw new IllegalArgumentException("placeholder is required");
        }
        if (name == null || name.isEmpty()) {
            throw new IllegalArgumentException("name is required");
        }
        return builder()
                .placeholder(placeholder)
                .name(name)
                .value(value)
                .mimeType(VariableMimeType.JSON)
                .usesAdvancedTemplatingEngine(true)
                .build();
    }

    /**
     * Creates a variable for array loops
     * @param placeholder The variable placeholder (e.g., "{products}")
     * @param name The variable name
     * @param value The array/list value
     * @throws IllegalArgumentException if any required parameter is missing
     */
    public static TemplateVariable loop(String placeholder, String name, List<?> value) {
        if (placeholder == null || placeholder.isEmpty()) {
            throw new IllegalArgumentException("placeholder is required");
        }
        if (name == null || name.isEmpty()) {
            throw new IllegalArgumentException("name is required");
        }
        return builder()
                .placeholder(placeholder)
                .name(name)
                .value(value)
                .mimeType(VariableMimeType.JSON)
                .usesAdvancedTemplatingEngine(true)
                .build();
    }

    /**
     * Creates a variable for conditionals
     * @param placeholder The variable placeholder (e.g., "{showDetails}")
     * @param name The variable name
     * @param value The boolean or truthy value
     * @throws IllegalArgumentException if any required parameter is missing
     */
    public static TemplateVariable conditional(String placeholder, String name, Object value) {
        if (placeholder == null || placeholder.isEmpty()) {
            throw new IllegalArgumentException("placeholder is required");
        }
        if (name == null || name.isEmpty()) {
            throw new IllegalArgumentException("name is required");
        }
        return builder()
                .placeholder(placeholder)
                .name(name)
                .value(value)
                .mimeType(VariableMimeType.JSON)
                .usesAdvancedTemplatingEngine(true)
                .build();
    }

    /**
     * Creates a variable for images
     * @param placeholder The variable placeholder (e.g., "{logo}")
     * @param name The variable name
     * @param imageUrl The image URL
     * @throws IllegalArgumentException if any required parameter is missing
     */
    public static TemplateVariable image(String placeholder, String name, String imageUrl) {
        if (placeholder == null || placeholder.isEmpty()) {
            throw new IllegalArgumentException("placeholder is required");
        }
        if (name == null || name.isEmpty()) {
            throw new IllegalArgumentException("name is required");
        }
        if (imageUrl == null || imageUrl.isEmpty()) {
            throw new IllegalArgumentException("imageUrl is required");
        }
        return builder()
                .placeholder(placeholder)
                .name(name)
                .value(imageUrl)
                .mimeType(VariableMimeType.IMAGE)
                .build();
    }
}

File: packages/php-sdk/src/Types/TemplateVariable.php
<?php

declare(strict_types=1);

namespace TurboDocx\Types;

/**
 * Variable configuration for template generation
 *
 * Supports both simple text replacement and advanced templating with Angular-like expressions
 */
class TemplateVariable
{
    /**
     * @param string $placeholder Variable placeholder (e.g., "{customer_name}")
     * @param string $name Variable name
     * @param VariableMimeType $mimeType MIME type of the variable
     * @param mixed $value Variable value (can be string, number, boolean, array, object, or null)
     * @param string|null $text Alternative to value for text content
     * @param bool|null $usesAdvancedTemplatingEngine Enable advanced templating engine
     * @param bool|null $nestedInAdvancedTemplatingEngine Marks variable as nested
     * @param bool|null $allowRichTextInjection Allow rich text injection
     * @param string|null $description Variable description
     * @param bool|null $defaultValue Whether this is a default value
     * @param array<TemplateVariable>|null $subvariables Nested sub-variables
     */
    public function __construct(
        public string $placeholder,
        public string $name,
        public VariableMimeType $mimeType,
        public mixed $value = null,
        public ?string $text = null,
        public ?bool $usesAdvancedTemplatingEngine = null,
        public ?bool $nestedInAdvancedTemplatingEngine = null,
        public ?bool $allowRichTextInjection = null,
        public ?string $description = null,
        public ?bool $defaultValue = null,
        public ?array $subvariables = null
    ) {
    }

    /**
     * Convert to array for JSON serialization
     *
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        $data = [
            'placeholder' => $this->placeholder,
            'name' => $this->name,
            'mimeType' => $this->mimeType->value,
        ];

        // Add value/text if present - allow null values
        if (array_key_exists('value', get_object_vars($this))) {
            $data['value'] = $this->value;
        }
        if ($this->text !== null) {
            $data['text'] = $this->text;
        }

        // Add optional boolean flags if set
        if ($this->usesAdvancedTemplatingEngine !== null) {
            $data['usesAdvancedTemplatingEngine'] = $this->usesAdvancedTemplatingEngine;
        }
        if ($this->nestedInAdvancedTemplatingEngine !== null) {
            $data['nestedInAdvancedTemplatingEngine'] = $this->nestedInAdvancedTemplatingEngine;
        }
        if ($this->allowRichTextInjection !== null) {
            $data['allowRichTextInjection'] = $this->allowRichTextInjection;
        }

        // Add optional string fields if set
        if ($this->description !== null) {
            $data['description'] = $this->description;
        }
        if ($this->defaultValue !== null) {
            $data['defaultValue'] = $this->defaultValue;
        }
        if ($this->subvariables !== null) {
            $data['subvariables'] = array_map(fn($v) => $v->toArray(), $this->subvariables);
        }

        return $data;
    }

    /**
     * Helper: Create a simple text variable
     *
     * @param string $placeholder Variable placeholder (e.g., "{customer_name}")
     * @param string $name Variable name
     * @param string|int|float|bool $value Variable value
     * @param VariableMimeType $mimeType MIME type ('TEXT' or 'HTML')
     * @return self
     */
    public static function simple(
        string $placeholder,
        string $name,
        string|int|float|bool $value,
        VariableMimeType $mimeType = VariableMimeType::TEXT
    ): self {
        if (empty($placeholder)) {
            throw new \InvalidArgumentException('placeholder is required');
        }
        if (empty($name)) {
            throw new \InvalidArgumentException('name is required');
        }
        if ($mimeType !== VariableMimeType::TEXT && $mimeType !== VariableMimeType::HTML) {
            throw new \InvalidArgumentException('mimeType must be TEXT or HTML for simple variables');
        }

        return new self(
            placeholder: $placeholder,
            name: $name,
            mimeType: $mimeType,
            value: $value
        );
    }

    /**
     * Helper: Create an advanced engine variable (for nested objects, complex data)
     *
     * @param string $placeholder Variable placeholder (e.g., "{user}")
     * @param string $name Variable name
     * @param array<string, mixed> $value Nested object value
     * @return self
     */
    public static function advancedEngine(
        string $placeholder,
        string $name,
        array $value
    ): self {
        if (empty($placeholder)) {
            throw new \InvalidArgumentException('placeholder is required');
        }
        if (empty($name)) {
            throw new \InvalidArgumentException('name is required');
        }

        return new self(
            placeholder: $placeholder,
            name: $name,
            mimeType: VariableMimeType::JSON,
            value: $value,
            usesAdvancedTemplatingEngine: true
        );
    }

    /**
     * Helper: Create a variable for array loops
     *
     * @param string $placeholder Variable placeholder (e.g., "{products}")
     * @param string $name Variable name
     * @param array<mixed> $value Array/list value
     * @return self
     */
    public static function loop(
        string $placeholder,
        string $name,
        array $value
    ): self {
        if (empty($placeholder)) {
            throw new \InvalidArgumentException('placeholder is required');
        }
        if (empty($name)) {
            throw new \InvalidArgumentException('name is required');
        }

        return new self(
            placeholder: $placeholder,
            name: $name,
            mimeType: VariableMimeType::JSON,
            value: $value,
            usesAdvancedTemplatingEngine: true
        );
    }

    /**
     * Helper: Create a variable for conditionals
     *
     * @param string $placeholder Variable placeholder (e.g., "{showDetails}")
     * @param string $name Variable name
     * @param mixed $value Boolean or truthy value
     * @return self
     */
    public static function conditional(
        string $placeholder,
        string $name,
        mixed $value
    ): self {
        if (empty($placeholder)) {
            throw new \InvalidArgumentException('placeholder is required');
        }
        if (empty($name)) {
            throw new \InvalidArgumentException('name is required');
        }

        return new self(
            placeholder: $placeholder,
            name: $name,
            mimeType: VariableMimeType::JSON,
            value: $value,
            usesAdvancedTemplatingEngine: true
        );
    }

    /**
     * Helper: Create a variable for images
     *
     * @param string $placeholder Variable placeholder (e.g., "{logo}")
     * @param string $name Variable name
     * @param string $imageUrl Image URL or base64 data
     * @return self
     */
    public static function image(
        string $placeholder,
        string $name,
        string $imageUrl
    ): self {
        if (empty($placeholder)) {
            throw new \InvalidArgumentException('placeholder is required');
        }
        if (empty($name)) {
            throw new \InvalidArgumentException('name is required');
        }
        if (empty($imageUrl)) {
            throw new \InvalidArgumentException('imageUrl is required');
        }

        return new self(
            placeholder: $placeholder,
            name: $name,
            mimeType: VariableMimeType::IMAGE,
            value: $imageUrl
        );
    }
}
